### ISSUE TITLE
Validate and upgrade the legacy WattLab dump into an ML-ready Open-FDD Engineering Data Bundle

### ISSUE THREAD

**Issue by bbartling** [bbartling/open-fdd#763]
## Summary

Open-FDD currently provides a versioned `wattlab_dump_v3` engineering bundle containing FDD findings, mapped telemetry, sensor statistics, schedules, weather data, equipment metadata, EnergyPlus model-seed inputs, and other RCx analytics.

The project also includes an offline `eplus_clustering_v1` exporter that creates:

* `clustering_features.csv`
* `clustering_features.parquet`
* `clustering_timeseries_long.parquet`
* `fdd_fault_vector.csv`
* `MANIFEST.json`
* An example KMeans workflow

This is a useful foundation, but it is not yet a complete, validated machine-learning handoff. The current clustering feature table contains approximately one statistical row per piece of equipment. It does not provide the time-windowed observations, target definitions, chronological train/validation/test splits, feature lineage, data-quality gates, or leakage controls needed for serious ML development.

This issue should validate the existing ZIP and evolve it into a portable, reproducible, AI-agent-friendly Open-FDD Engineering Data Bundle suitable for:

1. HVAC engineering and RCx analysis
2. EnergyPlus model development and calibration
3. Unsupervised clustering and anomaly detection
4. Supervised classification and regression experiments
5. Time-series forecasting
6. External AI-agent workflows

This issue is about producing a trustworthy dataset and engineering handoff. It is not a request to add model training or Python/sklearn execution to the Rust product image.

## Important current-state limitation

The React UI currently presents controls for building and downloading a WattLab ZIP, but `openfdd-central` disables this exporter by default because it depends on optional offline Python tooling.

The production architecture requires:

* Rust/DataFusion in the central product path
* No Python or sklearn dependency in the central/web images
* Optional Python tooling outside the product runtime
* No silent pandas fallback for production FDD calculations

The final design must preserve those boundaries.

## Phase 1 — Audit and validate the current export

Create an automated bundle validator and run it against:

* Synthetic-59
* Dirty-data fixtures
* A sanitized representative building fixture
* A local BUILDING_100 regression when that private dataset is available

The validator must confirm:

* [ ] The ZIP opens successfully.
* [ ] Every path listed in `MANIFEST.json` exists.
* [ ] Every required file is represented in the manifest.
* [ ] CSV and Parquet files can be read.
* [ ] Manifest-declared columns match the actual columns.
* [ ] Timestamps parse as timezone-aware UTC values.
* [ ] Equipment IDs and canonical point roles are stable.
* [ ] Units and equipment types are documented.
* [ ] No host-specific absolute paths are included.
* [ ] No credentials, tokens, private URLs, or environment values are included.
* [ ] Empty optional outputs are explicitly marked rather than silently omitted.
* [ ] Summary, diagnostic, and forensic profiles behave as documented.
* [ ] Package size, file count, row counts, and hashes are internally consistent.
* [ ] An external agent can understand the bundle without access to the Open-FDD source tree.

Provide a command such as:

```bash
openfdd-bundle validate openfdd_engineering_bundle.zip
```

The validator should return a machine-readable report:

```json
{
  "schema_version": "openfdd_bundle_validation_v1",
  "status": "READY_WITH_WARNINGS",
  "errors": [],
  "warnings": [],
  "files_checked": 0,
  "content_fingerprint": "...",
  "ml_readiness": {}
}
```

Supported statuses:

* `READY`
* `READY_WITH_WARNINGS`
* `NOT_READY`

## Phase 2 — Introduce a new versioned bundle contract

Replace the user-facing “WattLab dump” name with:

> Open-FDD Engineering & ML Data Bundle

Suggested schema:

```text
openfdd_engineering_bundle_v1
```

Keep legacy `/wattlab/dumps` routes temporarily as compatibility aliases, but introduce a neutral export route such as:

```text
POST /api/jobs/{job_id}/exports
```

The React UI should use language such as:

* Build Engineering & ML Bundle
* Download Engineering Bundle
* Export profile
* Include ML-ready datasets

Do not break existing `wattlab_dump_v3` consumers without an explicit compatibility and migration test.

## Phase 3 — Define the bundle structure

Suggested layout:

```text
MANIFEST.json
README.md

provenance/
  revisions.json
  checksums.sha256
  export_parameters.json

catalog/
  equipment.parquet
  points.parquet
  roles.parquet
  units.json

data/
  telemetry/
    <equipment_id>.parquet
  weather.parquet
  meters.parquet

quality/
  data_quality_summary.json
  coverage_by_point.parquet
  timestamp_gaps.parquet
  duplicate_timestamps.parquet
  validation_report.json

fdd/
  results.parquet
  findings.parquet
  fault_masks/
  rule_parameters.json

features/
  equipment_summary_features.parquet
  time_window_features.parquet
  feature_catalog.json

labels/
  labels.parquet
  label_catalog.json

splits/
  chronological_splits.json

energyplus/
  model_seed.json
  schedule_inference.json
  operating_signatures.parquet
  weather_observed.parquet
  missing_inputs.json

examples/
  README.md
  clustering_example.py
  forecasting_example.py
```

Small engineering summaries may remain CSV for accessibility, but canonical high-volume telemetry and ML tables should use Parquet.

### Canonical telemetry

Every telemetry record must have enough information to resolve:

* `timestamp_utc`
* Original site timezone
* `building_id`
* `equipment_id`
* `equipment_type`
* Canonical role
* Source point or source column
* Value
* Unit
* Data type
* Quality flag
* Mapping revision

Lossless, mapped telemetry should be included before any imputation or scaling.

Do not silently:

* Fill missing values
* Remove outliers
* Shuffle rows
* Resample data
* Normalize values
* Convert FDD results into ground-truth labels

Any transformation must be recorded in the manifest or feature catalog.

### Time-windowed features

Retain the current equipment-summary clustering features, but add a real time-windowed table with one row per:

```text
building × equipment × window_start
```

Potential features include:

* Count and coverage
* Mean, standard deviation, minimum, maximum, and percentiles
* Missing-value percentage
* Runtime fraction
* Setpoint error
* Rate of change
* Trend or slope
* Lagged values
* Rolling statistics
* Weather features
* Occupancy state
* Hour-of-day and day-of-week encodings
* FDD rule outputs when explicitly requested

Each feature must be documented in `feature_catalog.json` with:

* Feature name
* Source roles
* Window duration
* Formula or transformation
* Units
* Minimum required history
* Null behavior
* Feature-builder version
* Whether it could expose future information

### Labels and targets

Support multiple explicitly identified label sources:

* `fdd_weak_label`
* `manual_engineer_label`
* `synthetic_fault_label`
* `energyplus_simulation_target`
* `meter_target`

FDD findings must be described as weak or heuristic labels—not automatically treated as verified ground truth.

The label catalog must document:

* Target name
* Meaning
* Source
* Units
* Forecast horizon, if applicable
* Alignment method
* Class distribution
* Known limitations

### Time-aware splits

Generate deterministic chronological split boundaries rather than random row splits:

```json
{
  "strategy": "chronological",
  "train": {"start": "...", "end": "..."},
  "validation": {"start": "...", "end": "..."},
  "test": {"start": "...", "end": "..."},
  "gap_between_splits": "..."
}
```

Do not generate splits when the dataset is too short. Report that condition as an ML-readiness warning.

Features for a timestamp must use only information available at or before that timestamp unless explicitly classified as an offline retrospective feature.

## Phase 5 — Add comprehensive data-quality reporting

Calculate and report:

* Timestamp parsing success
* Duplicate timestamps
* Out-of-order timestamps
* Median polling interval
* Irregular sampling
* Long gaps
* Coverage by point and equipment
* Missing-value percentage
* Flat or low-variance points
* Out-of-range values
* Unit completeness
* Role-mapping completeness
* Equipment-type completeness
* Weather coverage
* Occupancy coverage
* Label availability
* Label imbalance
* Required history for selected feature windows

Dirty-data tests must include:

* Nulls
* Duplicate timestamps
* Out-of-order timestamps
* DST transitions
* Missing roles
* Mixed or incorrect units
* Constant sensors
* Short data windows
* Large gaps
* Multiple polling intervals
* Empty equipment
* Weather years that differ from telemetry or utility-bill years

Data-quality problems should not necessarily prevent an export, but the bundle must clearly distinguish warnings from contract-breaking errors.

## Phase 6 — Add provenance and reproducibility

The manifest must include:

* Exporter schema version
* Open-FDD version
* Git commit SHA
* Export timestamp
* Requested data window
* Building and job identifiers
* Source-data fingerprint
* Mapping revision/hash
* Equipment-model revision/hash
* Rule-registry hash
* Rule-parameter hash
* Unit-normalization version
* Feature-recipe version
* Per-file SHA-256
* Row count, column count, data types, and byte size per file

A reproducible export mode should produce a stable content fingerprint for identical data, configuration, and code. Volatile fields such as export time must not change the logical dataset fingerprint.

## Phase 7 — Provide ML smoke tests

Add lightweight tests proving the bundle is usable outside Open-FDD.

### Unsupervised smoke test

* Load `equipment_summary_features.parquet`
* Select numeric features
* Impute missing values through an explicit sklearn pipeline
* Scale values
* Run PCA and KMeans
* Produce one cluster assignment per equipment

### Time-series or supervised smoke test

* Load `time_window_features.parquet`
* Load a documented target
* Apply the provided chronological splits
* Train one simple baseline model
* Report MAE/RMSE for regression or precision/recall/F1 for classification
* Confirm that no random row shuffle occurs

These are contract tests and examples, not production ML-model claims.

## Phase 8 — Make large exports safe for Railway and low-memory hosts

The exporter must not read every file into memory before writing the ZIP.

Requirements:

* [ ] Stream files into the archive.
* [ ] Stream or batch DataFusion/Parquet output.
* [ ] Allow start/end time filters.
* [ ] Allow equipment and role filters.
* [ ] Estimate export size before generation.
* [ ] Apply configurable maximum ZIP and row limits.
* [ ] Provide a clear error when the requested export is too large.
* [ ] Avoid duplicating the entire historian inside job storage.
* [ ] Benchmark memory, runtime, compressed size, and uncompressed size.
* [ ] Clean up incomplete exports after failure.

A single ZIP should be treated as a bounded portable handoff, not as the storage format for a terabyte-scale historian. Larger deployments should export a manifest referencing partitioned Parquet datasets or object storage.

## Product architecture

The preferred architecture is:

```text
Rust/DataFusion product
    → canonical engineering bundle
    → downloadable ZIP or partitioned Parquet export
    → optional external Python/ML tooling
    → results attached back to an Open-FDD job
```

Do not add pandas, sklearn, or arbitrary Python execution to `openfdd-central`.

If advanced feature generation remains Python-based, it must run as:

* An offline CLI
* An optional separate worker/container
* A trusted external agent workflow

It must not become a hidden dependency of the default Rust product request path.

## Definition of done

* [ ] The default product can create a useful canonical bundle without Python.
* [ ] The UI no longer advertises an unavailable default exporter.
* [ ] Legacy WattLab consumers remain covered by compatibility tests.
* [ ] The new bundle has a documented, versioned schema.
* [ ] The bundle passes automated structural and data-quality validation.
* [ ] Parquet telemetry is lossless and preserves units and semantic roles.
* [ ] Time-windowed ML features have documented lineage.
* [ ] Labels and targets have explicit provenance.
* [ ] Chronological splits prevent obvious time leakage.
* [ ] KMeans and one time-aware baseline ML smoke test run successfully.
* [ ] No host-specific paths or secrets appear in the bundle.
* [ ] Export memory usage is bounded.
* [ ] Documentation explains what the bundle can and cannot prove.

## Related work

* #581 — Open-FDD/Vibe roadmap and job-native handoffs
* #211 — Future graph ML layers and ML Lab
* #104 — Earlier baked sklearn and YAML-recipe concept

This issue is scoped to the portable data/export contract. It should not revive obsolete Streamlit, PostgreSQL, or embedded-ML runtime architecture from older issues.

**Comment by 84dnnvbdvp-debug** [comment]
Current-source check found two concrete mismatches that seem worth folding into the Phase 1 validator/contract tests:

1. `scripts/eplus_dump_clustering_export.py` resolves `building_root` and then writes `building_root`, `artifact_root`, and the generated output file paths directly into `MANIFEST.json`. Those values are host-specific paths today, which conflicts with this issue's portability requirement.

2. The default `--max-long-rows 500000` path does `timeseries.sample(n=max_long_rows, random_state=0).sort_values("timestamp")`. So a large `clustering_timeseries_long.parquet` is a deterministic random subset of historian rows, not the full historian or a contiguous time window. The generated README still describes it as a “Long historian melt”, and the manifest records only the final `timeseries_rows`; it does not record the source row count, that sampling occurred, or the selection strategy/time range. The current agent-facing clustering doc also does not disclose this default reduction.

A small regression boundary could therefore be: require bundle-relative/logical manifest paths, and require any non-lossless row selection to be explicit and provenance-bearing (at minimum source/exported row counts plus selection strategy and relevant time range). Alternatively, make the default lossless and require an opt-in bounded export. For the time-series/forecasting use cases in this issue, I would avoid treating random-row capping as an invisible implementation detail.

**Comment by bbartling** [comment]
Thanks ill get this patched ASAP.

**Comment by 84dnnvbdvp-debug** [comment]
Correction to my follow-up: I over-read the surviving optional Python enrichment script as the canonical #763 export path.

Current `master` does in fact contain the Rust-first `openfdd_engineering_bundle_v1` implementation (`services/central/src/engineering_bundle.rs`), the `/api/jobs/{id}/exports` product path, a bundle validator, and docs that explicitly describe `scripts/eplus_dump_clustering_export.py` as optional offline enrichment while `/wattlab/dumps` is a deprecated alias.

So the two host-path/random-cap caveats I re-identified still exist in that optional offline script, but they do **not** establish that #763’s canonical implementation is missing. The completed closure state is consistent with the maintained product path. No reopen request from me.

**Comment by bbartling** [comment]
Agreed — thanks for the correction. Canonical #763 stays closed on the Rust `openfdd_engineering_bundle_v1` / `/api/jobs/{id}/exports` path. The two caveats in optional `scripts/eplus_dump_clustering_export.py` (host-absolute path insertion via repo-root `sys.path`, and random `DataFrame.sample` for the long-row cap) are offline-only and do not reopen the issue. We’ll track those as small optional hygiene separately; they are not part of the product bundle contract.
